zilencer/views.py

```diff
@@ -219,13 +220,68 @@ def unregister_all_remote_push_devices(
     return json_success(request)
 
 
+def delete_duplicate_registrations(server_id: int, user_id: int, user_uuid: str) -> None:
+    """
+    When migrating to support registration by UUID, we introduced a bug where duplicate
+    registrations for the same device+user could be created - one by user_id and one by
+    user_uuid. Given no good way of detecting these duplicates at database level, we need to
+    take advantage of the fact that when a remote server sends a push notification request
+    to us, it sends both user_id and user_uuid of the user.
+
+    This function, knowing the user_id and user_uuid of the user, can detect duplicates
+    and delete the legacy user_id registration if appropriate.
+    """
+
+    with connection.cursor() as cursor:
+        # We could write a more complicated, single query to detect and delete the duplicates
+        # in one query, but the cases where this function will proceed to having to do the additional
+        # delete query will be extremely rare, and thus are not a relevant concern.
+        query = SQL(
+            """
+            SELECT token, count(token) from zilencer_remotepushdevicetoken
+            WHERE server_id={server_id} AND (user_id={user_id} OR user_uuid={user_uuid})
+            GROUP BY token
+            HAVING count(token) > 1;
+            """
+        ).format(
+            server_id=Literal(server_id), user_id=Literal(user_id), user_uuid=Literal(user_uuid)
+        )
+        cursor.execute(query)
+        result = cursor.fetchall()
+
+    tokens_to_deduplicate = []
+    for token, count in result:
+        if count > 2:
+            raise AssertionError(
+                f"More than two registrations for token {token} for user id:{user_id} uuid:{user_uuid}, shouldn't be possible"
+            )
+        tokens_to_deduplicate.append(token)
+
+    if tokens_to_deduplicate:
+        logger.info(
+            "Deduplicating push registrations for server id:%s user id:%s uuid:%s and tokens:%s",
+            server_id,
+            user_id,
+            user_uuid,
+            sorted(tokens_to_deduplicate),
+        )
+        RemotePushDeviceToken.objects.filter(
+            token__in=tokens_to_deduplicate, server_id=server_id, user_id=user_id
+        ).delete()
```

Comment:
This function assumes that we  can ignore  token `kind` (android vs apple) because `token` on its own uniquely identifies a device and you can't have an apple device and an android device having identical tokens. I can tweak this if I'm wrong in this approach and we should be making sure to include `kind` in all such queries

Reply:
On the other hand, our unique indexes include kind - `unique_together = ("user", "kind", "token")`

Reply:
I think we should include `kind`. It's probably not important in that it's unlikely a Google token will have the same value as an apple one, but seems logically cleaner.

Reply:
And we definitely want this query to use an index, since we're doing it on every push notification request.

Reply:
That said ... since we're querying all of the devices already inside `remote_server_notify_push`, would it make sense to do this as Python logic using the ` android_devices` and `apple_devices` values rather than doing new queries? That would avoid any performance considerations in the common case where there's no duplicates, and might be about as easy to read.

Reply:
It's kinda uglier in python, but might be worth not having to worry about performance, yeah. I pushed   a `[squash]` commit with the rewrite